packages/@tailwindcss-upgrade/src/template/codemods/theme-to-var.ts

```diff
@@ -0,0 +1,237 @@
+    return substituteFunctionsInValue(input, (path, fallback) => {
+      let parts = segment(path, '/').map((part) => part.trim())
+
+      if (
+        // The path contains a `/`, which means that there is a modifier such as
+        // `theme(colors.red.500/50%)`. We can't convert this to a CSS variable.
+        parts.length > 1 ||
+        // In some scenarios (variants), we can't migrate to `var(…)` if it ends
+        // up in the `@media (…)` part. To be safe, let's just modernize the
+        // `theme(…)`
+        modernizeThemeOnly
+      ) {
+        let path = parts.shift()!
+
+        // Convert to modern `theme(…)` syntax using CSS variable syntax if we
+        // can.
+        let variable = `--${keyPathToCssProperty(toKeyPath(path))}` as const
+
+        if (!designSystem.theme.get([variable])) {
+          return null
+        }
+
+        let modifier = parts.length > 0 ? `/${parts.join('/')}` : ''
+
+        return fallback
+          ? `theme(${variable}${modifier}, ${fallback})`
+          : `theme(${variable}${modifier})`
+      }
+
+      // Convert to `var(…)`
+      let variable = `--${keyPathToCssProperty(toKeyPath(path))}` as const
+      if (!designSystem.theme.get([variable])) return null
+
+      return fallback ? `var(${variable}, ${fallback})` : `var(${variable})`
+    })
+  }
+
+  function convert(input: string, containsModifier = false) {
+    if (containsModifier) return [convertBasic(input), null] as const
+
+    let modifiers: CandidateModifier[] = []
+    let result = substituteFunctionsInValue(input, (path, fallback) => {
+      let parts = segment(path, '/').map((part) => part.trim())
+
+      // The path contains a `/`, which means that there is a modifier such as
+      // `theme(colors.red.500/50%)`.
+      //
+      // Currently, we are assuming that this is only being used for colors,
+      // which means that we can typically convert them to a modifier on the
+      // candidate itself.
+      if (parts.length === 2) {
+        let [pathPart, modifierPart] = parts
+
+        // 50% -> /50
+        if (/^\d+%$/.test(modifierPart)) {
+          modifiers.push({ kind: 'named', value: modifierPart.slice(0, -1) })
+        }
+
+        // .12 -> /12
+        // .12345 -> /[12.345]
+        else if (/^0?\.\d+$/.test(modifierPart)) {
+          let value = Number(modifierPart) * 100
+          modifiers.push({
+            kind: Number.isInteger(value) ? 'named' : 'arbitrary',
+            value: value.toString(),
+          })
+        }
+
+        // Anything else becomes arbitrary
+        else {
+          modifiers.push({ kind: 'arbitrary', value: modifierPart })
+        }
+
+        // Update path to be the first part
+        path = pathPart
+      }
+
+      let variable = `--${keyPathToCssProperty(toKeyPath(path))}` as const
+      if (!designSystem.theme.get([variable])) return null
+
+      return fallback ? `var(${variable}, ${fallback})` : `var(${variable})`
```

Comment:
I feel like there may be some design where this shared bits between convert and convertBasic aren't duplicated but perhaps the duplication is fine.

Reply:
Let me think about this one. It's definitely a little bit awkward because we are analyzing _while_ replacing. But depending on later information the final result might change.

At least all the necessary logic is there, but a clean 2 step pass might be enough.